File: mobile/lib/services/bank_service.dart
import 'package:dio/dio.dart';
import 'package:mobile/services/api_service.dart';

class BankService {
  /// Lấy snapshot số dư hiện tại
  static Future<Response> getSnapshot() async {
    return ApiService.client.get("/accountant/bank/snapshot");
  }

  /// Lịch sử giao dịch ngân hàng
  static Future<Response> getHistory({
    String? fromDate, // yyyy-MM-dd
    String? toDate,   // yyyy-MM-dd
    int page = 1,
    int size = 20,
  }) async {
    return ApiService.client.get(
      "/accountant/bank/history",
      queryParameters: {
        if (fromDate != null) "fromDate": fromDate,
        if (toDate != null) "toDate": toDate,
        "page": page,
        "size": size,
      },
    );
  }

  /// Refresh snapshot
  static Future<Response> refresh() async {
    return ApiService.client.post("/accountant/bank/refresh");
  }

  /// Lấy giao dịch thô từ PaymentController (/accountant/bank-transactions)
  static Future<Response> getRawTransactions({
    String? type, // CREDIT | DEBIT
    int page = 1,
    int size = 20,
  }) async {
    return ApiService.client.get(
      "/accountant/bank-transactions",
      queryParameters: {
        if (type != null) "type": type,
        "page": page,
        "size": size,
      },
    );
  }
}

File: client/src/services/bank.service.js
import api from "~/utils/axios";



// Nạp tiền
export const createTopupApi = async (data = {}) => {
  try {
    const res = await api.post("/payments/topups/bulk", data, {
      headers: { "Content-Type": "application/json" },
    });
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};

// Lịch sử topup của chính user
export const getMyTopupsApi = async ({ page = 1, size = 20, scope } = {}) => {
  try {
    const params = new URLSearchParams();
    params.set("page", page);
    params.set("size", size);
    if (scope) params.set("scope", scope); 
    const res = await api.get(`/payments/topups?${params.toString()}`);
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};

// Kiểm tra trạng thái nạp theo code
export const getTopupStatusApi = async (code) => {
  try {
    const safe = encodeURIComponent(code);
    const res = await api.get(`/payments/topups/status/${safe}`);
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};

// Lấy QR cho topup 
export const getTopupQrImageApi = async (code) => {
  try {
    const safe = encodeURIComponent(code);
    const res = await api.get(`/payments/topups/${safe}/qr`);
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};


// Helper format ngày YYYY-MM-DD theo BE
const fmtDate = (d) => {
  if (!d) return undefined;
  if (d instanceof Date) {
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, "0");
    const day = String(d.getDate()).padStart(2, "0");
    return `${y}-${m}-${day}`;
  }
  return String(d);
};

// Snapshot số dư hiện tại
export const getBankSnapshotApi = async () => {
  try {
    const res = await api.get("/accountant/bank/snapshot");
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};

// Lịch sử giao dịch 
export const getBankTransactionsPageApi = async (filter = {}) => {
  try {
    const {
      from,        
      to,          
      page = 1,    
      size = 20,
    } = filter;

    const params = new URLSearchParams();
    const fromDate = fmtDate(from);
    const toDate = fmtDate(to);
    if (fromDate) params.set("fromDate", fromDate);
    if (toDate)   params.set("toDate", toDate);
    params.set("page", page);
    params.set("size", size);

    const res = await api.get(`/accountant/bank/history?${params.toString()}`);
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};

// Đồng bộ lại số dư Fund 
export const refreshBankApi = async () => {
  try {
    const res = await api.post("/accountant/bank/refresh");
    return res.data;
  } catch (error) {
    return error.response?.data || { status: 500, message: "server-is-busy" };
  }
};
